Framework/Core/src/StepTHn.cxx

```diff
@@ -202,10 +210,24 @@ void StepTHn::Copy(TObject& c) const
   }
 
   if (mPrototype) {
-    target.mPrototype = dynamic_cast<THnSparseF*>(mPrototype->Clone());
+    if (mPrototype->InheritsFrom(THnSparseF::Class())) {
+      target.mPrototype = dynamic_cast<THnSparseF*>(mPrototype->Clone());
+    } else if (mPrototype->InheritsFrom(THnSparseD::Class())) {
+      target.mPrototype = dynamic_cast<THnSparseD*>(mPrototype->Clone());
+    }
   }
 }
 
+template <class TemplateArray>
+TObject* StepTHnT<TemplateArray>::Clone(const char* name)
+{
+  StepTHnT<TemplateArray>* newHist = new StepTHnT<TemplateArray>();
+  newHist->SetName((name) ? name : GetName());
+  newHist->SetTitle(GetTitle());
+  Copy(*newHist);
+  return newHist;
+}
```

Comment:
@jgrosseo please note that this new implementation of Clone() will also have side effects for the code in
`CorrelationContainer::Copy()`.
For instance in line `Analysis/Core/src/CorrelationContainer.cxx:460`
where `dynamic_cast<StepTHn*>mPairHist->Clone()` is called.
It may well be that I am misunderstanding how ROOT internally does the cloning (if so please correct me), but I would expect that the current implementation calls `TNamed::Clone()` and therefore the dynamic_cast to StepTHn always returns `nullptr`.
With my changes it should also there clone the actual StepTHn.
Please let me know in case this breaks any existing code.

Reply:
So ROOT::Clone() doesn't internally call ::Copy() on the actual instance?
If not you are right. Strange that this wasn't noticed.

Reply:
It may be that it does, one would have to check in the ROOT code. But I somehow cannot imagine how it would be technically possible for ROOT could call `new MyObject()` from within a Base-Class function without knowing the actual type of the object.

Reply:
It is indeed as you said. Seems like I underestimated ROOT ;)